ipysurfer/_utils/funcs.py

```diff
@@ -0,0 +1,203 @@
+
+    Parameters
+    ----------
+    subjects_dir : str | None
+        The subjects directory.
+    raise_error : bool
+        If True, raise a ValueError if no value for SUBJECTS_DIR can be found
+        or the corresponding directory does not exist.
+
+    Returns
+    -------
+    subjects_dir : str
+        The subjects directory. If the subjects_dir input parameter is not
+        None, its value will be returned, otherwise it will be obtained from
+        the SUBJECTS_DIR environment variable.
+    """
+    if subjects_dir is None:
+        subjects_dir = os.environ.get("SUBJECTS_DIR", "")
+        if not subjects_dir and raise_error:
+            raise ValueError('The subjects directory has to be specified '
+                             'using the subjects_dir parameter or the '
+                             'SUBJECTS_DIR environment variable.')
+
+    if raise_error and not os.path.exists(subjects_dir):
+        raise ValueError('The subjects directory %s does not exist.'
+                         % subjects_dir)
+
+    return subjects_dir
+
+
+def _mesh_edges(faces):
+    u"""Return sparse matrix with edges as an adjacency matrix.
+
+    Parameters
+    ----------
+    faces : array of shape [n_triangles x 3]
+        The mesh faces.
+
+    Returns
+    -------
+    edges : sparse matrix
+        The adjacency matrix.
+    """
+    npoints = np.max(faces) + 1
+    nfaces = len(faces)
+    a, b, c = faces.T
+    edges = sparse.coo_matrix((np.ones(nfaces), (a, b)),
+                              shape=(npoints, npoints))
+    edges = edges + sparse.coo_matrix((np.ones(nfaces), (b, c)),
+                                      shape=(npoints, npoints))
+    edges = edges + sparse.coo_matrix((np.ones(nfaces), (c, a)),
+                                      shape=(npoints, npoints))
+    edges = edges + edges.T
+    edges = edges.tocoo()
+
+    return edges
+
+
+def _smoothing_matrix(vertices, adj_mat, smoothing_steps=20, verbose=None):
+    """Create a smoothing matrix.
+
+    It can be used to interpolate data defined for a subset of
+    vertices onto mesh with an adjancency matrix given by
+    adj_mat.
+    If smoothing_steps is None, as many smoothing steps are applied until
+    the whole mesh is filled with with non-zeros. Only use this option if
+    the vertices correspond to a subsampled version of the mesh.
+
+    Parameters
+    ----------
+    vertices : 1d array
+        vertex indices.
+    adj_mat : sparse matrix
+        N x N adjacency matrix of the full mesh.
+    smoothing_steps : int or None
+        number of smoothing steps (Default: 20).
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see surfer.verbose).
+
+    Returns
+    -------
+    smooth_mat : sparse matrix
+        smoothing matrix with size N x len(vertices).
+    """
+    logger.info("Updating smoothing matrix, be patient..")
+
+    e = adj_mat.copy()
```

Comment:
something more expressive than "e". should be clear from the variable name what it is

Reply:
ok